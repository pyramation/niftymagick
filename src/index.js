export * from './assets';
export * from './klass';
export * from './prompter';
export * from './utils';
