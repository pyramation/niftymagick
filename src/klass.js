import { parse, resolve, basename, dirname, join } from 'path';
import * as crypto from 'crypto';
import stringify from 'safe-stable-stringify';
import { sync as mkdirp } from 'mkdirp';
import { writeFileSync } from 'fs';
import { ScriptMagick } from 'scriptmagick';

export class NiftyMagick {
  constructor({ traits, meta, outDir }) {
    this.meta = meta;
    this.traits = traits;
    this.hash = crypto
      .createHash('sha256')
      .update(stringify(traits))
      .digest('hex');
    if (!outDir) {
      this.outDir = join(process.cwd(), 'out', this.hash);
    } else {
      this.outDir = outDir;
    }
    this.scriptMagick = new ScriptMagick({ outDir: this.outDir });
    this.json = parse(this.hash).name + '.json';
    this.save = parse(this.hash).name + '.png';
    this.fileName = basename(this.save);
    this.outFile = resolve(join(this.outDir, this.fileName));
    this.outJson = resolve(join(this.outDir, this.json));
    this.layers = [];
    this.out = {};
    mkdirp(dirname(this.outFile));
  }
  push(pth) {
    this.layers.push(pth);
  }
  flatten() {
    this.scriptMagick.composeImages({
      layers: this.layers,
      outFile: this.outFile
    });
  }
  cleanup() {
    this.scriptMagick.cleanup();
  }
  writeJson() {
    writeFileSync(this.outJson, JSON.stringify(this.traits, null, 2));
  }
}
