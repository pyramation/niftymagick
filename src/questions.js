import inflection from 'inflection';

export const getQuestionsFromMeta = (schemas = [], meta) => {
  const exploded = Object.values(meta).reduce((m, v) => {
    return [...m, ...Object.values(v)];
  }, []);
  const questions = exploded.reduce((m, v) => {
    const qName = inflection.camelize('has_' + v.type, true);
    const schema = schemas.find((s) => s.type === v.type);
    if (schema && schema.ignore) return m;
    m[qName] = {
      type: 'confirm',
      name: qName,
      message: qName,
      requiredBy: [v.type]
    };
    return m;
  }, {});
  return questions;
};
