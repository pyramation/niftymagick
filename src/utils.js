// https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
export const getDeps = (deps) => {
  function depResolver(modulename, resolved, unresolved) {
    unresolved.push(modulename);
    if (!deps[modulename]) {
      throw new Error('no module exists: ' + modulename);
    }
    const edges = deps[modulename].requires || [];
    for (let i = 0; i < edges.length; i++) {
      const dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${modulename}, ${dep}`);
        }
        depResolver(dep, resolved, unresolved);
      }
    }
    resolved.push(modulename);
    const index = unresolved.indexOf(modulename);
    unresolved.splice(index);
  }

  return (name) => {
    const resolved = [];
    const unresolved = [];
    depResolver(name, resolved, unresolved);
    return resolved;
  };
};

export const convertQuestionsToDeps = (questions) => {
  return questions.reduce((m, v) => {
    m[v.name] = v;
    return m;
  }, {});
};

export const getQuestionOrder = (questions) => {
  const deps = convertQuestionsToDeps(questions);

  // require everything and then get the order...
  const getDependencies = getDeps({
    ...deps,
    root: {
      name: 'root',
      requires: Object.keys(deps)
    }
  });
  const order = getDependencies('root');
  order.pop();
  return order;
};
