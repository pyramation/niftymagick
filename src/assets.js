import { resolve, dirname, join } from 'path';
import { sync as glob } from 'glob';
import { readFileSync } from 'fs';
import inflection from 'inflection';
import { getQuestionOrder } from './utils';

const getFiles = (dir) => {
  if (!dir) return [];
  return glob(resolve(join(dir, '/**/*.json')));
};

export class NiftyAssets {
  constructor({ assetDir, schemaDir, additionalQuestions }) {
    this.content = [];
    this.assetFiles = getFiles(assetDir);
    this.schemaFiles = getFiles(schemaDir);
    this.additionalQuestions = additionalQuestions || [];
    this.processAssets();
    this.processSchemata();
    this.processConfirmations();
    this.processQuestions();
    this.processOptions();
    this.parseDeps();
  }
  processSchemata() {
    const objs = this.assetFiles.reduce((m, v) => {
      const content = JSON.parse(readFileSync(v, 'utf-8'));
      if (!content.type) {
        throw new Error(v + ' missing type:');
      }
      if (!content.name) {
        throw new Error(v + ' missing name:');
      }
      m[content.type] = content;
      return m;
    }, {});
    this.schemas = Object.values(objs);
    this.schemaHash = objs;
  }
  processAssets() {
    this.meta = this.assetFiles.reduce((m, v) => {
      const dir = dirname(v);
      const content = JSON.parse(readFileSync(v, 'utf-8'));
      if (!content.type) {
        throw new Error(v + ' missing type:');
      }
      if (!content.name) {
        throw new Error(v + ' missing name:');
      }
      content.dir = dir;
      m[content.type] = m[content.type] || {};
      m[content.type][content.name] = content;

      content.paths = {};
      if (content.image) content.paths.image = join(content.dir, content.image);
      if (content.fill) content.paths.fill = join(content.dir, content.fill);
      if (content.stroke)
        content.paths.stroke = join(content.dir, content.stroke);
      if (content.shadow)
        content.paths.shadow = join(content.dir, content.shadow);
      if (content.back) {
        content.paths.back = {};
        if (content.back.fill)
          content.paths.back.fill = join(content.dir, content.back.fill);
        if (content.back.stroke)
          content.paths.back.stroke = join(content.dir, content.back.stroke);
      }
      this.content.push(content);
      return m;
    }, {});
  }
  processConfirmations() {
    const confirms = this.content.reduce((m, v) => {
      const qName = inflection.camelize('has_' + v.type, true);
      const schema = this.schemas.find((s) => s.type === v.type);
      if ((schema && schema.ignore) || (schema && schema.required)) return m;
      m[qName] = {
        type: 'confirm',
        name: qName,
        message: qName,
        requiredBy: [v.type]
      };
      return m;
    }, {});
    this.confirmations = Object.values(confirms);
  }
  processOptions() {
    const singleOpts = {};
    const opts = Object.keys(this.meta).reduce((m, key) => {
      const keys = Object.keys(this.meta[key]);
      if (keys.length < 2) {
        singleOpts[key] = this.meta[key][keys[0]].name;
      }
      // ask user for which item of the possible types...
      return [
        ...m,
        {
          type: 'list',
          name: key,
          message: key,
          choices: keys
        }
      ];
    }, []);
    this.choices = opts;
    this.singleOpts = singleOpts;
    return opts;
  }
  processQuestions() {
    // update colors to be lists with choosing a bg color
    this.importedQuestions = this.additionalQuestions.map((q) => {
      switch (q.type) {
        case 'color':
          if (!q.meta) {
            throw new Error('colors require meta field');
          }
          if (!this.meta[q.meta]) {
            throw new Error('colors require meta field that exists');
          }
          return {
            name: q.name,
            message: q.name,
            ...q,
            type: 'list',
            choices: Object.keys(this.meta[q.meta])
          };
        default:
          return {
            name: q.name,
            message: q.name,
            ...q
          };
      }
    });
  }
  parseDeps() {
    const all = [
      ...this.choices,
      ...this.confirmations,
      ...this.importedQuestions
    ];
    const questionHash = all.reduce((m, v) => {
      m[v.name] = v;
      return m;
    }, {});
    this.questions = all.map((x) => {
      if (x.requiredBy) {
        /*
          {
            "type": "confirm",
            "name": "hasHat",
            "requiredBy": ["hat"]
          }
        */
        x.requiredBy.forEach((r) => {
          const obj = questionHash[r];
          obj.requires = obj.requires || [];
          if (!obj.requires.includes(x.name)) {
            obj.requires.push(x.name);
          }
          questionHash[r] = obj;
        });
      }
      if (x.excludedBy) {
        /*
          {
            "type": "confirm",
            "name": "otherhat",
            "excludedBy": ["hat"]
          }
        */
        x.excludedBy.forEach((r) => {
          const obj = questionHash[r];
          obj.excludes = obj.excludes || [];
          if (!obj.excludes.includes(x.name)) {
            obj.excludes.push(x.name);
          }
          questionHash[r] = obj;
        });
      }
    });
    const chk = () => true;
    const log = (t) => {
      // console.log(t);
    };
    const __questions = Object.values(questionHash).map((q) => {
      if (q.willExclude) {
        // write that check() to include
        Object.keys(q.willExclude).forEach((key) => {
          const val = q.willExclude[key];
          const that = questionHash[key];

          // update requires
          that.requires = that.requires || [];
          if (!that.requires.includes(q.name)) {
            that.requires.push(q.name);
          }

          log(`${q.name} if val:${val} excludes ${key}`);
          const check = !that.check ? chk : that.check;
          that.check = (traits) => {
            return traits[q.name] == val && check();
          };
        });
      }
      if (q.excludes && q.excludes.length) {
        // write that check() to include
        q.excludes.forEach((t) => {
          const that = questionHash[t];

          // update requires
          that.requires = that.requires || [];
          if (!that.requires.includes(q.name)) {
            that.requires.push(q.name);
          }
          if (
            q.type === 'confirm' &&
            // if it's not mentioned in exclusions
            (!q.exclusions || !q.exclusions.hasOwnProperty(t))
          ) {
            log(`${q.name} if true excludes ${t}`);
            const check = !that.check ? chk : that.check;
            that.check = (traits) => {
              return traits[q.name] && check();
            };
          }
        });
      }
      if (q.requires && q.requires.length) {
        // write this check() to include
        q.requires.forEach((t) => {
          const that = questionHash[t];
          if (
            that.type === 'confirm' &&
            // if it's not mentioned in includedWhen
            (!q.includedWhen || !q.includedWhen.hasOwnProperty(t))
          ) {
            log(`${q.name} requires ${t} is true`);
            const check = !q.check ? chk : q.check;
            q.check = (traits) => {
              return traits[q.name] && check();
            };
          }
        });
      }
      if (q.includedWhen) {
        // write this check() to include
        Object.keys(q.includedWhen).forEach((key) => {
          const val = q.includedWhen[key];
          const that = questionHash[key];

          // update requires
          q.requires = q.requires || [];
          if (!q.requires.includes(key)) {
            q.requires.push(key);
          }

          log(`${q.name} includedWhen ${key} is val:${val}`);
          const check = !q.check ? chk : q.check;
          q.check = (traits) => {
            return traits[q.name] == val && check();
          };
        });
      }
      if (q.excludedWhen) {
        // write this check() to include
        Object.keys(q.excludedWhen).forEach((key) => {
          const val = q.excludedWhen[key];
          const that = questionHash[key];

          // update requires
          q.requires = q.requires || [];
          if (!q.requires.includes(key)) {
            q.requires.push(key);
          }

          log(`${q.name} excludedWhen ${key} is val:${val}`);
          const check = !q.check ? chk : q.check;
          q.check = (traits) => {
            return traits[q.name] != val && check();
          };
        });
      }
      return q;
    });

    this.questionOrder = getQuestionOrder(__questions);
    this.questions = getQuestionOrder(__questions).map((name) => {
      return __questions.find((a) => a.name === name);
    });
  }
}
