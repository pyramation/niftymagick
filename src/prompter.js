import { prompt as inquire } from 'inquirerer';
import { NiftyAssets } from './assets';

export class TraitPrompter {
  constructor({ argv, assetDir, schemaDir, additionalQuestions }) {
    this.argv = argv;

    // assets
    this.assets = new NiftyAssets({
      assetDir,
      schemaDir,
      additionalQuestions
    });
    this.questions = this.assets.questions;
  }
  async getTraits() {
    const argv = this.argv;
    for (let i = 0; i < this.questions.length; i++) {
      if (this.questions[i].check) {
        if (this.questions[i].check(this.traits)) {
          this.traits = await inquire([this.questions[i]], {
            ...argv,
            ...this.singleOpts,
            ...this.traits
          });
        }
      } else {
        this.traits = await inquire([this.questions[i]], {
          ...argv,
          ...this.singleOpts,
          ...this.traits
        });
      }
    }
    return this.traits;
  }
  async ask(questions) {
    const argv = this.argv;
    return await inquire(questions, { ...argv, ...this.singleOpts });
  }
}
