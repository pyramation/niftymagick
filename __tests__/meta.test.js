import { NiftyAssets, getQuestionsFromMeta } from '../src';
// assets
const assets = new NiftyAssets({
  assetDir: __dirname + '/../__fixtures__/assets',
  schemaDir: __dirname + '/../__fixtures__/schemas',
  additionalQuestions: [
    {
      type: 'color',
      name: 'hairColor',
      meta: 'pattern',
      requires: ['hasHair'],
      includedWhen: {
        hasHair: true
      }
    },
    {
      type: 'confirm',
      name: 'hasHairPattern',
      requires: ['hasHair'],
      excludes: ['hairColor'],
      willExclude: {
        hairColor: true
      }
    },
    {
      type: 'confirm',
      name: 'anotherThing',
      excludedWhen: {
        hasHair: false,
        hasHairPattern: true
      }
    },
    {
      type: 'color',
      name: 'hairPattern',
      meta: 'pattern',
      requires: ['hasHairPattern'],
      includedWhen: {
        hasHairPattern: true
      }
    }
  ]
});
it('confirmations', async () => {
  expect(JSON.stringify(assets.confirmations, null, 2)).toMatchSnapshot();
});
it('choices', async () => {
  expect(JSON.stringify(assets.choices, null, 2)).toMatchSnapshot();
});
it('singleOpts', async () => {
  expect(JSON.stringify(assets.singleOpts, null, 2)).toMatchSnapshot();
});
it('all', async () => {
  expect(JSON.stringify(assets.questions, null, 2)).toMatchSnapshot();
});
it('schema', async () => {
  expect(JSON.stringify(assets.schemas, null, 2)).toMatchSnapshot();
});
