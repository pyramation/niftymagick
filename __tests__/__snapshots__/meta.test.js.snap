// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all 1`] = `
"[
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasAnother\\",
    \\"message\\": \\"hasAnother\\",
    \\"requiredBy\\": [
      \\"another\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"another\\",
    \\"message\\": \\"another\\",
    \\"choices\\": [
      \\"another1\\",
      \\"another2\\"
    ],
    \\"requires\\": [
      \\"hasAnother\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasHair\\",
    \\"message\\": \\"hasHair\\",
    \\"requiredBy\\": [
      \\"hair\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"hair\\",
    \\"message\\": \\"hair\\",
    \\"choices\\": [
      \\"hair1\\"
    ],
    \\"requires\\": [
      \\"hasHair\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasHat\\",
    \\"message\\": \\"hasHat\\",
    \\"requiredBy\\": [
      \\"hat\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"hat\\",
    \\"message\\": \\"hat\\",
    \\"choices\\": [
      \\"hat1\\"
    ],
    \\"requires\\": [
      \\"hasHat\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasOtherhat\\",
    \\"message\\": \\"hasOtherhat\\",
    \\"requiredBy\\": [
      \\"otherhat\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"otherhat\\",
    \\"message\\": \\"otherhat\\",
    \\"choices\\": [
      \\"otherhat1\\"
    ],
    \\"requires\\": [
      \\"hasOtherhat\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasPattern\\",
    \\"message\\": \\"hasPattern\\",
    \\"requiredBy\\": [
      \\"pattern\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"pattern\\",
    \\"message\\": \\"pattern\\",
    \\"choices\\": [
      \\"pattern1\\"
    ],
    \\"requires\\": [
      \\"hasPattern\\"
    ]
  },
  {
    \\"name\\": \\"hasHairPattern\\",
    \\"message\\": \\"hasHairPattern\\",
    \\"type\\": \\"confirm\\",
    \\"requires\\": [
      \\"hasHair\\"
    ],
    \\"excludes\\": [
      \\"hairColor\\"
    ],
    \\"willExclude\\": {
      \\"hairColor\\": true
    }
  },
  {
    \\"name\\": \\"hairColor\\",
    \\"message\\": \\"hairColor\\",
    \\"type\\": \\"list\\",
    \\"meta\\": \\"pattern\\",
    \\"requires\\": [
      \\"hasHair\\",
      \\"hasHairPattern\\"
    ],
    \\"includedWhen\\": {
      \\"hasHair\\": true
    },
    \\"choices\\": [
      \\"pattern1\\"
    ]
  },
  {
    \\"name\\": \\"anotherThing\\",
    \\"message\\": \\"anotherThing\\",
    \\"type\\": \\"confirm\\",
    \\"excludedWhen\\": {
      \\"hasHair\\": false,
      \\"hasHairPattern\\": true
    },
    \\"requires\\": [
      \\"hasHair\\",
      \\"hasHairPattern\\"
    ]
  },
  {
    \\"name\\": \\"hairPattern\\",
    \\"message\\": \\"hairPattern\\",
    \\"type\\": \\"list\\",
    \\"meta\\": \\"pattern\\",
    \\"requires\\": [
      \\"hasHairPattern\\"
    ],
    \\"includedWhen\\": {
      \\"hasHairPattern\\": true
    },
    \\"choices\\": [
      \\"pattern1\\"
    ]
  }
]"
`;

exports[`choices 1`] = `
"[
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"another\\",
    \\"message\\": \\"another\\",
    \\"choices\\": [
      \\"another1\\",
      \\"another2\\"
    ],
    \\"requires\\": [
      \\"hasAnother\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"hair\\",
    \\"message\\": \\"hair\\",
    \\"choices\\": [
      \\"hair1\\"
    ],
    \\"requires\\": [
      \\"hasHair\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"hat\\",
    \\"message\\": \\"hat\\",
    \\"choices\\": [
      \\"hat1\\"
    ],
    \\"requires\\": [
      \\"hasHat\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"otherhat\\",
    \\"message\\": \\"otherhat\\",
    \\"choices\\": [
      \\"otherhat1\\"
    ],
    \\"requires\\": [
      \\"hasOtherhat\\"
    ]
  },
  {
    \\"type\\": \\"list\\",
    \\"name\\": \\"pattern\\",
    \\"message\\": \\"pattern\\",
    \\"choices\\": [
      \\"pattern1\\"
    ],
    \\"requires\\": [
      \\"hasPattern\\"
    ]
  }
]"
`;

exports[`confirmations 1`] = `
"[
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasAnother\\",
    \\"message\\": \\"hasAnother\\",
    \\"requiredBy\\": [
      \\"another\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasHair\\",
    \\"message\\": \\"hasHair\\",
    \\"requiredBy\\": [
      \\"hair\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasHat\\",
    \\"message\\": \\"hasHat\\",
    \\"requiredBy\\": [
      \\"hat\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasOtherhat\\",
    \\"message\\": \\"hasOtherhat\\",
    \\"requiredBy\\": [
      \\"otherhat\\"
    ]
  },
  {
    \\"type\\": \\"confirm\\",
    \\"name\\": \\"hasPattern\\",
    \\"message\\": \\"hasPattern\\",
    \\"requiredBy\\": [
      \\"pattern\\"
    ]
  }
]"
`;

exports[`singleOpts 1`] = `
"{
  \\"hair\\": \\"hair1\\",
  \\"hat\\": \\"hat1\\",
  \\"otherhat\\": \\"otherhat1\\",
  \\"pattern\\": \\"pattern1\\"
}"
`;
